### ISSUE TITLE
more efficient monitor calculation

### ISSUE THREAD

**Issue by erik-roberts** [DynaSim/DynaSim#69]
monitors seem to be just passive and not related to the numerical solving.  it seems faster and more efficient to calculate them after the solution, since it would then be vectorized.  this could be implemented when disk_flag=0, i.e. not saving each step to disk.

**Comment by jsherfey** [comment]
Perhaps, counterintuitively, the proposed change is less efficient.

Using vectorized computation after simulation takes twice as long in this example:

>> tic; for i=1:1e7, x(i)=i; y(i)=i; end; z=x.*y; toc
Elapsed time is 0.146478 seconds.
>> tic; for i=1:1e7, x(i)=i; y(i)=i; z(i)=i.*i; end; toc
Elapsed time is 0.070124 seconds.

**Comment by kupiqu** [comment]
At the very least, this should be made dependent of matlab vs octave. The reason why it's faster in matlab (at least in this simple example) is because the use of JIT compiler in matlab.

Also, it is not trivial how this scales from that simple example to the more sophisticated solver files that we simulate...

**Comment by kupiqu** [comment]
I quantified how SERIOUS this issue is for Octave:

There is an additional problem in the simple example, because it does not set the size for x, y (and z in the second example) in advance.

This is the result in Octave after fixing for this:

```
tic; l = 1e7; x = nan(1,l); y = x; z = x; for i=1:l, x(i)=i; y(i)=i; end; z=x.*y; toc
Elapsed time is 70.1571 seconds.

tic; l = 1e7; x = nan(1,l); y = x; z = x; for i=1:l, x(i)=i; y(i)=i; z(i) = i.*i; end; toc
Elapsed time is 125.378 seconds.
```

That is ~55 more seconds in this simple example!! (I also tested using x = nan(l,1) and the example took even longer to run).

So, this is indeed sth that could speed up the simulation A LOT in Octave (while, most likely, matlan users would barely notice any difference) !!

**Comment by kupiqu** [comment]
Jason, it may be twice the time in matlab, but that's a difference of 0.07 seconds for a 1e7 iteration loop, while it's also almost twice the difference in the reverse direction for octave, but represents 55 seconds!

That's why I say that matlab users won't notice the difference, while this represents a huge speed up for octave users...

**Comment by kupiqu** [comment]
I made an attempt to vectorize monitors in Octave, but it's more tricky than expected. Jason, I'd like to discuss this with you...

**Comment by erik-roberts** [comment]
With a more realistic example, the vectorized version is definitely faster:
```
vlength = 1e8;

x1 = zeros(1, vlength);
x2 = zeros(1, vlength);
y1 = zeros(1, vlength);
y2 = zeros(1, vlength);
z1 = zeros(1, vlength);
% z2 = zeros(1, vlength);

disp('Vector')
tic;
for k=1:vlength 
  x2(k) = k^2;
  y2(k) = exp(k);
end
z2 = x2.*y2;
toc

disp('Non vector')
tic;
for i=1:vlength 
  x1(i) = i^2;
  y1(i) = exp(i);
%   z1(i) = x1(i) * y1(i);
  z1(i) = i^2 * exp(i);
end
toc
```
```
Vector
Elapsed time is 1.461184 seconds.
Non vector
Elapsed time is 1.843295 seconds.
```

in your example, there was no preallocation and shared variables, so whatever is run first is much slower.

**Comment by erik-roberts** [comment]
what was the status with this? i think i'm going to take a crack at it. @kupiqu do you have any notes on how far you got with it?

**Comment by kupiqu** [comment]
It's done already, only that it's only applied when using Octave. If we agree on using this in general (I don't see why not if it is in general faster), then we just need to remove the conditional for Octave ;-)

**Comment by erik-roberts** [comment]
nice! have you tested that it works in matlab, not just benchmark?

**Comment by kupiqu** [comment]
It would be good to test this in different models, before making a final decision. For this, you could fool DynaSim, setting the Octave conditional to either 0 or 1...

**Comment by erik-roberts** [comment]
good idea. ill work on it soon.  do you think it should go in dev or debug?

**Comment by kupiqu** [comment]
Right now this works in Matlab (monitors saved at each step) and Octave (Monitors saved at the end). This is stable, as far as I could test...

Now I recommend that you fool DynaSim, check it out etc and see what's the result.

I would start in dev, and then we see...

**Comment by jsherfey** [comment]
@kupiqu
<http://t.sidekickopen69.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJW7t5XX48p-YszW64QZjY2zGJWPW7fRL1x56dR7Sf83FkD602?t=https%3A%2F%2Fgithub.com%2Fkupiqu&si=4782755602235392&pi=848c62f1-87b4-468d-aafc-ec64f44924e1>
didn't
you discover limitations to the vectorized calculation that were not issues
with the non-vectorized approach? do you remember what the limitations are?
we should be careful not to break anything or reduce functionality in the
matlab version without a way to use the more capable approach. perhaps we
should keep things as they are but simply change the name of the flag
controlling the conditional from octave_flag to something more descriptive
like vectorize_monitors_flag.

**Comment by kupiqu** [comment]
As far as I remember the only limitation is that monitor expressions should follow matlab notation, e.g.: 

`'monitor w(t)=u.*v'                    % defining a function of state variables to monitor (note that monitor expressions follow Matlab's syntax)`

There was some discussion about random values such in:

`'I(t)=Iapp*(t>ton&t<toff).*(1+.5*rand(1,Npop))' % define applied input using reserved variables 't' for time and 'dt' for fixed time step of numerical`

but I don't see how this is any different as the code is vectorized in Npop anyways.

**Comment by kupiqu** [comment]
I agree on keeping/renaming the flag to vectorize_monitors_flag, but I don't think the limitation of the monitor expressions (following matlab notation) is enough to set its default value to 0 if it is proven that real models get significantly faster if vectorized...

**Comment by DynaSim** [comment]
I agree if it's proven to be faster.
